include/boost/compute/algorithm/detail/find_extrema_reduce.hpp

```diff
@@ -0,0 +1,245 @@
+                 "block_idx[lid] = compare_result ? " <<
+                     "block_idx[lid] : block_idx[lid+offset];\n" <<
+             "}\n"
+             "barrier(CLK_LOCAL_MEM_FENCE);\n" <<
+        "}\n" <<
+
+         // write block result to global output
+        "if(lid == 0){\n" <<
+        "    output[get_group_id(0)] = block[0];\n" <<
+        "    output_idx[get_group_id(0)] = block_idx[0];\n" <<
+        "}";
+
+    kernel kernel = k.compile(context);
+    kernel.set_arg(count_arg, static_cast<uint_>(count));
+    kernel.set_arg(output_arg, result.get_buffer());
+    kernel.set_arg(output_idx_arg, result_idx.get_buffer());
+    kernel.set_arg(block_arg, local_buffer<input_type>(work_group_size));
+    kernel.set_arg(block_idx_arg, local_buffer<uint_>(work_group_size));
+
+    queue.enqueue_1d_range_kernel(kernel,
+                                  0,
+                                  work_groups_no * work_group_size,
+                                  work_group_size);
+
+    return 0;
+}
+
+template<class InputIterator>
+InputIterator find_extrema_reduce(InputIterator first,
+                                  InputIterator last,
+                                  size_t preferred_work_group_size,
+                                  char sign,
+                                  command_queue &queue)
+{
+    typedef typename std::iterator_traits<InputIterator>::difference_type difference_type;
+    typedef typename std::iterator_traits<InputIterator>::value_type input_type;
+
+    const context &context = queue.get_context();
+    const device &device = queue.get_device();
+
+    // Getting information about used queue and device
+    const size_t compute_units_no = device.get_info<CL_DEVICE_MAX_COMPUTE_UNITS>();
+    const size_t max_work_group_size = device.get_info<CL_DEVICE_MAX_WORK_GROUP_SIZE>();
+    const bool is_queue_ordered =
+        static_cast<bool>(
+            queue.get_properties() &
+            command_queue::enable_out_of_order_execution
+        );
+
+    size_t count = detail::iterator_range_size(first, last);
+    size_t work_group_size = (std::min)(max_work_group_size, preferred_work_group_size);
+    size_t work_groups_no = compute_units_no * 64;
+    work_groups_no = (std::min)(
+            work_groups_no,
+            static_cast<size_t>(std::ceil(float(count) / work_group_size)));
+
+    // Device vectors for extremum candidates and their indices
+    vector<input_type> results(work_groups_no, context);
+    vector<uint_> results_idx(work_groups_no, context);
+
+    // Find extremum candidates and their indices
+    find_extrema_reduce(first, count,
+                        results.begin(), results_idx.begin(),
+                        work_groups_no, work_group_size,
+                        sign,
+                        queue);
+
+    std::vector<input_type> host_results(work_groups_no);
+    std::vector<uint_> host_results_idx(work_groups_no);
+
+    if(!is_queue_ordered) {
```

Comment:
Yes, you're right. I know ```event```-based synchronization should be used with out-of-order queue. I just thought "oh, it will not work with out-of-order queue" and I can't get ```event```, so I did what I did :) If we assume in-order, it's unnecessary perform ```finish()``` and indeed it affects other kernels. I will remove this part.